queue/queue.go

```diff
@@ -117,3 +122,36 @@ func (queue *Queue) PersistedIndex() (EntryID, error) {
 		return EntryID(metrics.OldestEntryID), nil
 	}
 }
+
+// WrappedBatch is a bookkeeping wrapper around a libbeat queue batch,
+// to work around the fact that shipper acknowledgements are per-event
+// while the queue can only track an entire batch at a time.
+// The plan is to eliminate WrappedBatch once batch assembly / acknowledgment
+// is moved out of the libbeat queue.
+type WrappedBatch struct {
+	batch beatsqueue.Batch
+
+	// how many events from the batch have been acknowledged
+	doneCount uint64
+
+	// If CompletionCallback is non-nil, wrappedBatch will call it
+	// when all events have been consumed.
+	CompletionCallback func()
+}
+
+func (w *WrappedBatch) Events() []*messages.Event {
+	events := make([]*messages.Event, w.batch.Count())
+	for i := 0; i < w.batch.Count(); i++ {
+		events[i], _ = w.batch.Entry(i).(*messages.Event)
+	}
+	return events
+}
+
+func (w *WrappedBatch) Done(count int) {
+	if atomic.AddUint64(&w.doneCount, uint64(count)) >= uint64(w.batch.Count()) {
```

Comment:
It is my fervent hope that WrappedBatch will not live long enough to cause problems since it's just a workaround until we deal with https://github.com/elastic/elastic-agent-shipper/issues/186, but I've made this parameter unsigned and pushed the conversion back to the call sites.